flocker/control/test/test_persistence.py

```diff
@@ -495,32 +447,62 @@ def _build_node(applications):
 SUPPORTED_VERSIONS = st.integers(1, _CONFIG_VERSION)
 
 
-class ConfigurationMigrationTests(SynchronousTestCase):
+class WireEncodeDecodeTests(SynchronousTestCase):
     """
-    Tests for ``ConfigurationMigration`` class that performs individual
-    configuration upgrades.
-
-    There should be
+    Tests for ``wire_encode`` and ``wire_decode``.
     """
+    def test_encode_to_bytes(self):
+        """
+        ``wire_encode`` converts the given object to ``bytes``.
+        """
+        self.assertIsInstance(wire_encode(TEST_DEPLOYMENT), bytes)
+
     @given(DEPLOYMENTS)
-    def test_upgrade_configuration_v1_latest(self, deployment):
+    def test_roundtrip(self, deployment):
         """
         A range of generated configurations (deployments) can be
-        upgraded from v1 to the latest configuration version.
-
-        This test will need updating for each new configuration version
-        introduced to reflect the expected configuration format
-        (``expected_configuration``) after a successful upgrade.
+        roundtripped via the wire encode/decode.
         """
         source_json = wire_encode(deployment)
-        upgraded_json = migrate_configuration(
-            1, _CONFIG_VERSION, source_json, ConfigurationMigration)
-        upgraded_config = wire_decode(upgraded_json)
-        expected_configuration = Configuration(
-            version=_CONFIG_VERSION, deployment=deployment
-        )
-        self.assertEqual(upgraded_config, expected_configuration)
+        decoded_deployment = wire_decode(source_json)
+        self.assertEqual(decoded_deployment, deployment)
+
+    def test_no_arbitrary_decoding(self):
+        """
+        ``wire_decode`` will not decode classes that are not in
+        ``SERIALIZABLE_CLASSES``.
+        """
+        class Temp(PRecord):
+            """A class."""
+        SERIALIZABLE_CLASSES.append(Temp)
 
+        def cleanup():
+            if Temp in SERIALIZABLE_CLASSES:
+                SERIALIZABLE_CLASSES.remove(Temp)
+        self.addCleanup(cleanup)
+
+        data = wire_encode(Temp())
+        SERIALIZABLE_CLASSES.remove(Temp)
+        # Possibly future versions might throw exception, the key point is
+        # that the returned object is not a Temp instance.
+        self.assertFalse(isinstance(wire_decode(data), Temp))
+
+    def test_complex_keys(self):
+        """
+        Objects with attributes that are ``PMap``\s with complex keys
+        (i.e. not strings) can be roundtripped.
+        """
+        node_state = NodeState(hostname=u'127.0.0.1', uuid=uuid4(),
+                               manifestations={}, paths={},
+                               devices={uuid4(): FilePath(b"/tmp")})
```

Comment:
No action suggested.

One of the nice things with hypothesis is that you can easily use it to fill in data in a specific case like this, e.g.

```python
node_state = NodeState(hostname=IP_ADDRESSES.example(), ...)
```

(Which is the [Anonymous Creation Method](http://xunitpatterns.com/Creation%20Method.html) pattern, for those playing at home)

But I don't think we have generators for host names or filepaths, and the uuid generator is exactly the same.